File: 提交版/最终提交/matlab/data_loader.m
function [stock_data, stock_codes] = load_all_stocks(data_dir, max_stocks, verbose)
% LOAD_ALL_STOCKS 批量加载 data_dir 下所有股票 CSV 文件
%
%   参数:
%       data_dir:   CSV 文件所在目录路径
%       max_stocks: 最大加载数量，[] 或省略表示全部
%       verbose:    true/false，是否打印进度（默认 true）
%
%   返回:
%       stock_data:  struct 数组，每个元素包含 .code 和 .data (table)
%       stock_codes: 字符串数组，股票代码列表

if nargin < 3, verbose = true; end
if nargin < 2, max_stocks = []; end

if ~isfolder(data_dir)
    error('数据目录不存在: %s', data_dir);
end

% 查找所有 CSV 文件
csv_files = dir(fullfile(data_dir, '*.csv'));
if isempty(csv_files)
    error('在 %s 中未找到 CSV 文件', data_dir);
end

if verbose
    fprintf('找到 %d 个 CSV 文件\n', length(csv_files));
end

% 限制数量
if ~isempty(max_stocks)
    csv_files = csv_files(1:min(max_stocks, length(csv_files)));
    if verbose
        fprintf('限制加载前 %d 只股票\n', max_stocks);
    end
end

% 初始化
n_files = length(csv_files);
stock_data = struct('code', cell(n_files, 1), 'data', cell(n_files, 1));
stock_codes = strings(n_files, 1);
valid_count = 0;
fail_count = 0;
skip_count = 0;

for i = 1:n_files
    filepath = fullfile(csv_files(i).folder, csv_files(i).name);

    try
        tbl = load_single_stock(filepath);
    catch ME
        if verbose
            fprintf('  [警告] 读取失败: %s — %s\n', csv_files(i).name, ME.message);
        end
        fail_count = fail_count + 1;
        continue;
    end

    if isempty(tbl) || height(tbl) < 30
        skip_count = skip_count + 1;
        continue;
    end

    valid_count = valid_count + 1;
    stock_data(valid_count).code = char(tbl.stock_code(1));
    stock_data(valid_count).data = tbl;
    stock_codes(valid_count) = string(tbl.stock_code(1));

    % 进度显示
    if verbose && mod(i, 500) == 0
        fprintf('  已处理 %d/%d 个文件...\n', i, n_files);
    end
end

% 截取有效部分
stock_data = stock_data(1:valid_count);
stock_codes = stock_codes(1:valid_count);

if verbose
    fprintf('加载完成: 成功 %d 只, 数据不足/跳过 %d 只, 失败 %d 只\n', ...
        valid_count, skip_count, fail_count);
end

end


function tbl = load_single_stock(filepath)
% LOAD_SINGLE_STOCK 读取单只股票的 CSV 文件，返回清洗后的 table
%
%   参数:
%       filepath: CSV 文件路径
%
%   返回:
%       table，包含标准化后的日线数据；若读取失败返回空

% 尝试读取 CSV
try
    tbl = readtable(filepath, 'VariableNamingRule', 'preserve', ...
        'TextType', 'string');
catch
    tbl = [];
    return;
end

% 空文件检查
if isempty(tbl) || height(tbl) == 0
    return;
end

% ---- 字段标准化 ----
% 获取列名
colnames = tbl.Properties.VariableNames;

% 检查必要字段
required_cols = {'day', 'close', 'open', 'high', 'low', 'volume'};
for c = 1:length(required_cols)
    if ~any(strcmp(colnames, required_cols{c}))
        warning('  [警告] %s 缺少字段: %s', filepath, required_cols{c});
        tbl = [];
        return;
    end
end

% 检查 hsl（换手率）字段
if ~any(strcmp(colnames, 'hsl'))
    warning('  [警告] 缺少 hsl 字段: %s', filepath);
    tbl = [];
    return;
end

% ---- 日期处理 ----
if iscell(tbl.day), tbl.day = string(tbl.day); end
tbl.day = datetime(tbl.day, 'InputFormat', 'yyyy-MM-dd', 'Locale', 'en_US');
% 移除无效日期
tbl(isnat(tbl.day), :) = [];
% 按日期排序
tbl = sortrows(tbl, 'day');

% ---- 数值列处理 ----
numeric_cols = {'open', 'close', 'high', 'low', 'volume', 'amount', ...
                'zf', 'zdf', 'zde', 'hsl'};
for c = 1:length(numeric_cols)
    col = numeric_cols{c};
    if any(strcmp(colnames, col))
        % 转换为数值
        if iscell(tbl.(col)) || isstring(tbl.(col))
            tbl.(col) = str2double(string(tbl.(col)));
        end
        % 确保为 double
        tbl.(col) = double(tbl.(col));
    end
end

% ---- 缺失值处理 ----
% 价格类字段：前向/后向填充
price_cols = {'open', 'close', 'high', 'low'};
for c = 1:length(price_cols)
    col = price_cols{c};
    if any(strcmp(colnames, col))
        tbl.(col) = fillmissing(tbl.(col), 'previous');
        tbl.(col) = fillmissing(tbl.(col), 'next');
    end
end

% 成交量/成交额：前向/后向填充
if any(strcmp(colnames, 'volume'))
    tbl.volume = fillmissing(tbl.volume, 'previous');
    tbl.volume = fillmissing(tbl.volume, 'next');
end
if any(strcmp(colnames, 'amount'))
    tbl.amount = fillmissing(tbl.amount, 'previous');
    tbl.amount = fillmissing(tbl.amount, 'next');
end

% ---- 涨跌幅重算验证 ----
if any(strcmp(colnames, 'zdf'))
    nan_ratio = sum(isnan(tbl.zdf)) / height(tbl);
    if nan_ratio > 0.5
        close_vals = tbl.close;
        zdf_vals = [NaN; (close_vals(2:end) - close_vals(1:end-1)) ./ close_vals(1:end-1) * 100];
        tbl.zdf = round(zdf_vals, 2);
    end
else
    close_vals = tbl.close;
    zdf_vals = [NaN; (close_vals(2:end) - close_vals(1:end-1)) ./ close_vals(1:end-1) * 100];
    tbl.zdf = round(zdf_vals, 2);
end

% ---- 振幅重算 ----
if any(strcmp(colnames, 'zf'))
    nan_ratio = sum(isnan(tbl.zf)) / height(tbl);
    if nan_ratio > 0.3
        prev_close = [tbl.close(1); tbl.close(1:end-1)];
        tbl.zf = round((tbl.high - tbl.low) ./ prev_close * 100, 2);
    end
else
    prev_close = [tbl.close(1); tbl.close(1:end-1)];
    tbl.zf = round((tbl.high - tbl.low) ./ prev_close * 100, 2);
end

% ---- 提取股票代码 ----
if any(strcmp(colnames, 'code'))
    stock_code = char(tbl.code(1));
else
    [~, name, ~] = fileparts(filepath);
    stock_code = name;
end

% 添加 stock_code 列（作为 categorical 或 cellstr）
tbl.stock_code = repmat({stock_code}, height(tbl), 1);

end


function summary = get_data_summary(stock_data)
% GET_DATA_SUMMARY 生成数据概览
%
%   参数:
%       stock_data: struct 数组，load_all_stocks 的返回结果
%
%   返回:
%       table: 每只股票的基本信息

n = length(stock_data);
codes = cell(n, 1);
rows = zeros(n, 1);
start_dates = NaT(n, 1);
end_dates = NaT(n, 1);
has_hsl = false(n, 1);
avg_volumes = zeros(n, 1);
avg_zdfs = zeros(n, 1);

for i = 1:n
    codes{i} = stock_data(i).code;
    tbl = stock_data(i).data;
    rows(i) = height(tbl);
    start_dates(i) = min(tbl.day);
    end_dates(i) = max(tbl.day);
    has_hsl(i) = any(strcmp(tbl.Properties.VariableNames, 'hsl'));
    avg_volumes(i) = mean(tbl.volume, 'omitnan');
    if any(strcmp(tbl.Properties.VariableNames, 'zdf'))
        avg_zdfs(i) = mean(tbl.zdf, 'omitnan');
    else
        avg_zdfs(i) = NaN;
    end
end

summary = table(string(codes), rows, start_dates, end_dates, has_hsl, ...
    avg_volumes, avg_zdfs, ...
    'VariableNames', {'stock_code', 'rows', 'start_date', 'end_date', ...
    'has_hsl', 'avg_volume', 'avg_zdf'});

end

File: 提交版/最终提交/stock_strategy/data_loader.py
# -*- coding: utf-8 -*-
"""
数据读取与预处理模块
- 遍历 data_dir 下所有 CSV 文件
- 读取单只股票数据，统一字段格式
- 缺失值处理、日期排序
"""

import os
import glob
import pandas as pd
import numpy as np
from typing import Dict, Tuple, Optional


def load_single_stock(filepath: str) -> Optional[pd.DataFrame]:
    """
    读取单只股票的 CSV 文件，返回清洗后的 DataFrame。

    参数:
        filepath: CSV 文件路径

    返回:
        DataFrame，包含标准化后的日线数据；若读取失败返回 None

    数据字段:
        day, code, open, close, high, low, volume, amount, zf, zdf, zde, hsl, market
    """
    try:
        df = pd.read_csv(filepath, encoding="utf-8")
    except (UnicodeDecodeError, FileNotFoundError) as e:
        print(f"  [警告] 读取失败: {filepath} — {e}")
        return None

    # 空文件检查
    if df.empty or len(df) == 0:
        return None

    # ---- 字段标准化 ----
    # 确保关键字段存在
    required_cols = ["day", "close", "open", "high", "low", "volume"]
    missing = [c for c in required_cols if c not in df.columns]
    if missing:
        print(f"  [警告] {filepath} 缺少字段: {missing}")
        return None

    # 检查 hsl（换手率）字段
    if "hsl" not in df.columns:
        print(f"  [警告] 缺少 hsl 字段: {filepath}")
        return None

    # ---- 日期处理 ----
    df["day"] = pd.to_datetime(df["day"], errors="coerce")
    df.dropna(subset=["day"], inplace=True)
    df.sort_values("day", inplace=True)
    df.reset_index(drop=True, inplace=True)

    # ---- 数值列处理 ----
    numeric_cols = ["open", "close", "high", "low", "volume", "amount",
                    "zf", "zdf", "zde", "hsl"]
    for col in numeric_cols:
        if col in df.columns:
            # 强制转换为数值类型，非数值设为 NaN
            df[col] = pd.to_numeric(df[col], errors="coerce")

    # ---- 缺失值处理 ----
    # 涨跌幅类字段：首个交易日 NaN 正常（无前一日比较），保留 NaN
    # 价格类字段：如果缺失则前向填充
    price_cols = ["open", "close", "high", "low"]
    for col in price_cols:
        if col in df.columns:
            # 仅填充内部的 NaN，首行 NaN 若存在则用下一有效值回填
            df[col] = df[col].ffill().bfill()

    # 成交量/成交额：前向填充
    df["volume"] = df["volume"].ffill().bfill()
    if "amount" in df.columns:
        df["amount"] = df["amount"].ffill().bfill()

    # ---- 涨跌幅重算验证 ----
    # 若原始 zdf 缺失过多，则根据收盘价重新计算
    if "zdf" in df.columns:
        nan_ratio = df["zdf"].isna().mean()
        if nan_ratio > 0.5:
            df["zdf"] = (df["close"].pct_change() * 100).round(2)
    else:
        df["zdf"] = (df["close"].pct_change() * 100).round(2)

    # ---- 振幅重算 ----
    if "zf" in df.columns:
        nan_ratio = df["zf"].isna().mean()
        if nan_ratio > 0.3:
            prev_close = df["close"].shift(1)
            df["zf"] = ((df["high"] - df["low"]) / prev_close * 100).round(2)
    else:
        prev_close = df["close"].shift(1)
        df["zf"] = ((df["high"] - df["low"]) / prev_close * 100).round(2)

    # ---- 提取股票代码 ----
    if "code" in df.columns:
        # 统一格式：取原始字符串，去除可能的空白
        stock_code = str(df["code"].iloc[0]).strip()
    else:
        # 从文件名推断
        basename = os.path.splitext(os.path.basename(filepath))[0]
        stock_code = basename

    df["stock_code"] = stock_code

    # ---- 检查最低数据量 ----
    if len(df) < 30:
        print(f"  [跳过] {stock_code}: 数据量不足 ({len(df)} 行)")
        return None

    return df


def load_all_stocks(data_dir: str, max_stocks: int = None) -> Dict[str, pd.DataFrame]:
    """
    批量加载 data_dir 下所有股票 CSV 文件。

    参数:
        data_dir: CSV 文件所在目录
        max_stocks: 最大加载数量，None 表示全部

    返回:
        dict: {股票代码: DataFrame}
    """
    if not os.path.isdir(data_dir):
        raise FileNotFoundError(f"数据目录不存在: {data_dir}")

    csv_files = sorted(glob.glob(os.path.join(data_dir, "*.csv")))
    if not csv_files:
        raise FileNotFoundError(f"在 {data_dir} 中未找到 CSV 文件")

    print(f"找到 {len(csv_files)} 个 CSV 文件")

    if max_stocks:
        csv_files = csv_files[:max_stocks]
        print(f"限制加载前 {max_stocks} 只股票")

    stock_data: Dict[str, pd.DataFrame] = {}
    skip_count = 0
    fail_count = 0

    for i, filepath in enumerate(csv_files):
        df = load_single_stock(filepath)
        if df is None or df.empty:
            fail_count += 1
            continue

        code = df["stock_code"].iloc[0]
        if len(df) < 30:
            skip_count += 1
            continue

        stock_data[code] = df

        # 进度显示
        if (i + 1) % 500 == 0:
            print(f"  已处理 {i + 1}/{len(csv_files)} 个文件...")

    print(f"加载完成: 成功 {len(stock_data)} 只, 数据不足/跳过 {skip_count} 只, "
          f"失败 {fail_count} 只")
    return stock_data


def get_data_summary(stock_data: Dict[str, pd.DataFrame]) -> pd.DataFrame:
    """
    生成数据概览，包含每只股票的基本信息。

    返回:
        DataFrame: 列包含 stock_code, rows, start_date, end_date, has_hsl, ...
    """
    records = []
    for code, df in stock_data.items():
        records.append({
            "stock_code": code,
            "rows": len(df),
            "start_date": df["day"].min(),
            "end_date": df["day"].max(),
            "has_hsl": "hsl" in df.columns,
            "avg_volume": df["volume"].mean(),
            "avg_zdf": df["zdf"].mean() if "zdf" in df.columns else np.nan,
        })
    return pd.DataFrame(records)
